### PR TITLE
fix(mongo): retry only 3 times if ensuring indexes fails

### LINKED ISSUE #3261: MongoDb: ensureIndexes() goes into infinite loop, when there is an error while creating indexes
https://github.com/mikro-orm/mikro-orm/blob/932a6e4a51178bc60db61281e44f93e16a84f137/packages/mongodb/src/MongoSchemaGenerator.ts#L77-L108

This error could be replicated by adding some entities with duplicate values onto the database, and then adding the @Unique() index on the property.

Database reports the error and promise becomes rejected, but nothing is logged by the ORM, and as it is recursive call if it gets failed, it keeps on looping.

Solution could look like this..

```ts
async ensureIndexes(options: EnsureIndexesOptions = {}, retryLimit = 3): Promise<void> { 
   .....
   if(retryLimit === 0) {
       // Crash with the error ?
   } 
   if (res.some(r => r.status === 'rejected') && options.retry !== false) { 
     const skipIndexes = res.filter(r => r.status === 'fulfilled').map((r: Dictionary, id) => ({ collection: promises[id][0], indexName: r.value })); 
     const collectionsWithFailedIndexes = res.filter(r => r.status === 'rejected').map((r: Dictionary, id) => promises[id][0]); 
     await this.dropIndexes({ skipIndexes, collectionsWithFailedIndexes }); 
     // Maybe some timeout before trying again ?
     await this.ensureIndexes({ retry: collectionsWithFailedIndexes }, retryLimit--); 
   } 
 } 
```

### PR DESCRIPTION
**Pull request by jagzmz**
Closes #3261

@B4nan I am not sure how would I reproduce the error on tests, they are not complete as of now, any suggestions ?

### CODE REVIEW

**Comment by B4nan** [comment]
Here it should be enough if you define the unique constraint, but dont create it right ahead, create the two rows, and then call `ensureIndexes()` afterwards. First comment out the changes you made to verify such test will fail, then see how your fix is working.

File: tests/features/schema-generator/GH3261.test.ts [hunk 1]
```diff
@@ -0,0 +1,46 @@
+import { Collection, Entity, ManyToOne, MikroORM, OneToMany, PrimaryKey, Property, Unique } from '@mikro-orm/core';
+
+@Entity()
+export class User {
+
+  @PrimaryKey()
+  _id!: number;
+
+
+  @Property({ type: 'boolean' })
```
Comment (B4nan) [inline tests/features/schema-generator/GH3261.test.ts:10]:
```suggestion
  @Property()
```

File: tests/features/schema-generator/GH3261.test.ts [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+  email!: string;
+
+}
+
+
```
Comment (B4nan) [inline tests/features/schema-generator/GH3261.test.ts:15]:
```suggestion

```

File: tests/features/schema-generator/GH3261.test.ts [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+let orm: MikroORM;
+
+beforeAll(async () => {
+  orm = await MikroORM.init({
+    entities: [User],
+    dbName: 'GH3216',
```
Comment (B4nan) [inline tests/features/schema-generator/GH3261.test.ts:21]:
```suggestion
    dbName: 'mikro-orm-test-gh3261',
```

File: tests/features/schema-generator/GH3261.test.ts [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+
+afterAll(() => orm.close(true));
+
+test('retry limit to 3 when ensureIndex() fails', async () => {
+  const user1 = await orm.em.create(User, {
+    email: 'test',
+  });
+  const user2 = await orm.em.create(User, {
+    email: 'test',
+  });
```
Comment (B4nan) [inline tests/features/schema-generator/GH3261.test.ts:35]: `em.create` is sync method

```suggestion
  const user1 = orm.em.create(User, {
    email: 'test',
  });
  const user2 = orm.em.create(User, {
    email: 'test',
  });
```

**Review by B4nan — APPROVED** [review]
thanks! just few nits before we merge

File: packages/mongodb/src/MongoSchemaGenerator.ts [hunk 5]
```diff
@@ -74,7 +74,7 @@ export class MongoSchemaGenerator extends AbstractSchemaGenerator<MongoDriver> {
     await Promise.all(promises);
   }
 
-  async ensureIndexes(options: EnsureIndexesOptions = {}): Promise<void> {
+  async ensureIndexes(options: EnsureIndexesOptions = {}, retryLimit = 3): Promise<void> {
```
Comment (B4nan) [inline packages/mongodb/src/MongoSchemaGenerator.ts:77]: actually, lets not introduce another parameter, we already have the options there, so lets just add the `retryLimit` inside that
```suggestion
  async ensureIndexes(options: EnsureIndexesOptions = {}): Promise<void> {
    options.retryLimit ??= 3
```

File: packages/mongodb/src/MongoSchemaGenerator.ts [hunk 6]
```diff
@@ -103,7 +103,11 @@ export class MongoSchemaGenerator extends AbstractSchemaGenerator<MongoDriver> {
       const skipIndexes = res.filter(r => r.status === 'fulfilled').map((r: Dictionary, id) => ({ collection: promises[id][0], indexName: r.value }));
       const collectionsWithFailedIndexes = res.filter(r => r.status === 'rejected').map((r: Dictionary, id) => promises[id][0]);
       await this.dropIndexes({ skipIndexes, collectionsWithFailedIndexes });
-      await this.ensureIndexes({ retry: collectionsWithFailedIndexes });
+      if (retryLimit === 1) {
+        const failedIndexes = res.filter(r => r.status === 'rejected').map((r: Dictionary, id) => `${promises[id][0]} - ${r.reason}`);
+        throw new Error(`Failed to create indexes: ${failedIndexes.join(', ')}`);
+      }
+      await this.ensureIndexes({ retry: collectionsWithFailedIndexes }, retryLimit - 1);
```
Comment (B4nan) [inline packages/mongodb/src/MongoSchemaGenerator.ts:110]:
```suggestion

      if (options.retryLimit === 1) {
        const failedIndexes = res.filter(r => r.status === 'rejected').map((r: Dictionary, id) => `${promises[id][0]} - ${r.reason}`);
        throw new Error(`Failed to create indexes: ${failedIndexes.join(', ')}`);
      }

      await this.ensureIndexes({
        retry: collectionsWithFailedIndexes,
        retryLimit: retryLimit - 1,
      });
```

**Review by B4nan — CHANGES_REQUESTED** [review]

File: packages/mongodb/src/MongoSchemaGenerator.ts (hunk 5 above)
Comment (jagzmz) [reply packages/mongodb/src/MongoSchemaGenerator.ts]: Updated!

File: packages/mongodb/src/MongoSchemaGenerator.ts (hunk 6 above)
Comment (jagzmz) [reply packages/mongodb/src/MongoSchemaGenerator.ts]: Updated

File: packages/mongodb/src/MongoSchemaGenerator.ts [hunk 7]
```diff
@@ -103,11 +104,14 @@ export class MongoSchemaGenerator extends AbstractSchemaGenerator<MongoDriver> {
       const skipIndexes = res.filter(r => r.status === 'fulfilled').map((r: Dictionary, id) => ({ collection: promises[id][0], indexName: r.value }));
       const collectionsWithFailedIndexes = res.filter(r => r.status === 'rejected').map((r: Dictionary, id) => promises[id][0]);
       await this.dropIndexes({ skipIndexes, collectionsWithFailedIndexes });
-      if (retryLimit === 1) {
+      if (options.retryLimit === 1) {
```
Comment (B4nan) [inline packages/mongodb/src/MongoSchemaGenerator.ts:107]:
```suggestion

      if (options.retryLimit === 0) {
```

File: packages/mongodb/src/MongoSchemaGenerator.ts [hunk 8]
```diff
@@ -103,11 +104,14 @@ export class MongoSchemaGenerator extends AbstractSchemaGenerator<MongoDriver> {
...
         const failedIndexes = res.filter(r => r.status === 'rejected').map((r: Dictionary, id) => `${promises[id][0]} - ${r.reason}`);
         throw new Error(`Failed to create indexes: ${failedIndexes.join(', ')}`);
       }
-      await this.ensureIndexes({ retry: collectionsWithFailedIndexes }, retryLimit - 1);
+      await this.ensureIndexes({
```
Comment (B4nan) [inline packages/mongodb/src/MongoSchemaGenerator.ts:111]:
```suggestion

      await this.ensureIndexes({
```

**Review by B4nan — APPROVED** [review]

**Comment by B4nan** [comment]
Thanks!

**Pull request merged**